### PR TITLE
Stability file

### LINKED ISSUE #100: Add support for stability configuration file for the Gradle analyzer
From what I can see, stability configuration file is currently only used in the IntelliJ plugin. It would be nice if it was also supported in the Gradle analyzer plugin.

### PR DESCRIPTION
**Pull request by matejdro**
### 🎯 Goal

This fixes #100. 

It also updates `NewFunction` message to include a list of all unstable parameters in this function to aid user with debugging (before, user had to manually regenerate the baseline or manually check, which parameter could be unstable).

This is my last PR in a batch. Any chance for a release when it gets merged?

### 🛠 Implementation details

PR copies over `getCustomStableTypesAsRegex()` from the IntelliJ project (as far as I can see, there is no good way to share the function?) and adds a gradle plugin option to specify the stability file. Then it uses aforementioned method to create a list of regexes and uses these regexes in the `compareStability()` (which has also been moved into its own file, so test do not contain a copy).

## Summary by CodeRabbit

* **New Features**
  * Specify multiple stability configuration files; checks apply user patterns.
  * Composable diffs include parameter-level stability details for new functions.

* **Documentation**
  * README example shows how to add stability configuration files.

* **Tests**
  * Expanded tests covering stability comparison, pattern parsing/matching, wildcards, regression filtering, and parameter-level scenarios.

### CODE REVIEW

File: stability-gradle/src/main/kotlin/com/skydoves/compose/stability/gradle/StabilityComparison.kt [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    }
+  }
+
+  return differences
+}
+
+private fun StabilityEntry.isStable(forceStableTypes: List<Regex>): Boolean {
+  return skippable ||
+    (parameters.isNotEmpty() && parameters.all { it.isStable(forceStableTypes) })
+}
```
Comment (skydoves) [inline stability-gradle/src/main/kotlin/com/skydoves/compose/stability/gradle/StabilityComparison.kt]: A composable with zero parameters would return false if not already marked skippable. Zero-parameter composables should generally be considered stable. Should be:

```kotlin
  private fun StabilityEntry.isStable(forceStableTypes: List<Regex>): Boolean {
    return skippable ||
      parameters.isEmpty() ||  // No params = stable
      parameters.all { it.isStable(forceStableTypes) }
  }
```

File: stability-gradle/src/main/kotlin/com/skydoves/compose/stability/gradle/StabilityCheckTask.kt [hunk 2]
```diff
@@ -466,80 +475,43 @@ public abstract class StabilityCheckTask : DefaultTask() {
...
-            )
+
+        // Convert patterns to regex
+        patterns.mapNotNull { pattern ->
+          try {
+            // Convert glob-style wildcards to regex
+            pattern
+              .replace(".", "\\.")
+              .replace("*", ".*")
+              .toRegex()
```
Comment (skydoves) [inline stability-gradle/src/main/kotlin/com/skydoves/compose/stability/gradle/StabilityCheckTask.kt]: So here, ** is incorrectly handled. For example, `com.example.**` becomes `com\.example\..*.*` (two .* in sequence) instead of `com\.example\..*` Also, we're missing `?` support,  which is a single character wildcard is common in globs.

Would consider:

```kotlin
  patterns.mapNotNull { pattern ->
    try {
      // Escape all regex metacharacters first, then convert glob wildcards
      val escaped = Regex.escape(pattern)
      // Convert glob wildcards (after escaping, * becomes \*, so we look for that)
      escaped
        .replace("\\*\\*", ".*")  // ** -> match anything including dots
        .replace("\\*", "[^.]*")   // * -> match anything except dots (single segment)
        .replace("\\?", ".")       // ? -> single character
        .toRegex()
    } catch (e: Exception) {
      logger.warn("Invalid stability pattern: $pattern")
      null
    }
  }
```

File: stability-gradle/src/main/kotlin/com/skydoves/compose/stability/gradle/StabilityComparison.kt [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+private fun ParameterInfo.isStable(forceStableTypes: List<Regex>): Boolean {
+  return stability == "STABLE" || forceStableTypes.any { it.matches(type) }
+}
```
Comment (skydoves) [inline stability-gradle/src/main/kotlin/com/skydoves/compose/stability/gradle/StabilityComparison.kt]: `matches()` requires the entire string to match. If type is `List<com.example.MyType>`, a pattern like `com.example.*` won't match because it only matches the inner type portion. We should consider utilizing `containsMatchIn(type)` for partial matching.

File: stability-gradle/src/main/kotlin/com/skydoves/compose/stability/gradle/StabilityCheckTask.kt [hunk 4]
```diff
@@ -90,6 +91,9 @@ public abstract class StabilityCheckTask : DefaultTask() {
   @get:Input
   public abstract val allowMissingBaseline: Property<Boolean>
 
+  @get:InputFiles
+  public abstract val stabilityConfigurationFiles: ListProperty<RegularFile>
+
```
Comment (skydoves) [inline stability-gradle/src/main/kotlin/com/skydoves/compose/stability/gradle/StabilityCheckTask.kt:101]: `@InputFiles` should specify path sensitivity. we can add `@get:PathSensitive(PathSensitivity.RELATIVE)` here.

File: stability-gradle/src/main/kotlin/com/skydoves/compose/stability/gradle/StabilityCheckTask.kt [hunk 5]
```diff
@@ -549,8 +521,15 @@ public abstract class StabilityCheckTask : DefaultTask() {
...
-    override fun format(): String = "+ $name (new composable)"
+  public data class NewFunction(
+    val name: String,
+    val parameters: List<ParameterInfo>,
+  ) : StabilityDifference() {
+    override fun format(): String {
+      val endColon = if (parameters.isEmpty()) "" else ":"
+      return "+ $name (new composable)$endColon\n" +
+        "${parameters.joinToString("\n") { "    ${it.name}: ${it.stability}" }}"
+    }
```
Comment (skydoves) [inline stability-gradle/src/main/kotlin/com/skydoves/compose/stability/gradle/StabilityCheckTask.kt:513]:
```suggestion
override fun format(): String {
    return if (parameters.isEmpty()) {
      "+ $name (new composable)"
    } else {
      "+ $name (new composable):\n" +
        parameters.joinToString("\n") { "    ${it.name}: ${it.stability}" }
    }
  }
```

File: stability-gradle/src/test/kotlin/com/skydoves/compose/stability/gradle/StabilityComparisonTest.kt [hunk 6]
```diff
@@ -69,6 +69,66 @@ class StabilityComparisonTest {
     assertEquals("com.example.New2", (differences[0] as StabilityDifference.NewFunction).name)
   }
 
+  @Test
+  fun testCompareStability_newFunctionWithRegressionFilteringAndStableRegex() {
+    val current = mapOf(
```
Comment (skydoves) [inline stability-gradle/src/test/kotlin/com/skydoves/compose/stability/gradle/StabilityComparisonTest.kt:74]: It would be good to have some missing unit tests:

  - Tests for invalid/missing configuration file paths
  - Tests for malformed configuration files (e.g., binary files, encoding issues)
  - Tests for ** glob pattern edge case
  - Tests for regex metacharacters in patterns (e.g., pattern containing + or ()

File: stability-gradle/src/main/kotlin/com/skydoves/compose/stability/gradle/StabilityComparison.kt (hunk 1 above)
Comment (matejdro) [reply stability-gradle/src/main/kotlin/com/skydoves/compose/stability/gradle/StabilityComparison.kt]: Hm, My original logic is to use the "skippable" property if there are no parameters.

Does that mean that we can actually ignore the `skippable` property? Or is there a case where a function would be skippable, but  its parameters are not stable?

File: stability-gradle/src/main/kotlin/com/skydoves/compose/stability/gradle/StabilityComparison.kt (hunk 3 above)
Comment (matejdro) [reply stability-gradle/src/main/kotlin/com/skydoves/compose/stability/gradle/StabilityComparison.kt]: I think that's what it should do.

 `List<com.example.MyType>` should be matched by a `List`, not `com.example.MyType` (~~I don't think Google's compose stability checker even does generics~~).

Partial matching would cause matches where they shouldn't be. For example `mypackage.**` would also match `anotherpackage.mypackage.SomeClass`, which it shouldn't as `anotherpackage` is not on the list and there is no glob for matching it.

File: stability-gradle/src/main/kotlin/com/skydoves/compose/stability/gradle/StabilityCheckTask.kt (hunk 2 above)
Comment (matejdro) [reply stability-gradle/src/main/kotlin/com/skydoves/compose/stability/gradle/StabilityCheckTask.kt]: From [Google's doc](https://developer.android.com/develop/ui/compose/performance/stability/fix#configuration-file):

> Comments, single, and double wildcards are supported

So I don't think `?` is actually supported?

Comment (matejdro) [reply stability-gradle/src/main/kotlin/com/skydoves/compose/stability/gradle/StabilityCheckTask.kt]: I started with this, but after looking at the generics syntax, I just gave up and copied over Google's parser

File: stability-gradle/src/test/kotlin/com/skydoves/compose/stability/gradle/StabilityComparisonTest.kt (hunk 6 above)
Comment (matejdro) [reply stability-gradle/src/test/kotlin/com/skydoves/compose/stability/gradle/StabilityComparisonTest.kt:74]: Added

**Comment by matejdro** [comment]
@skydoves I have looked at generics and I don't think approach in this PR will work at all.

For example, if I have this class:

`class MyGenericClass<T>`

by default, it will be unstable. However, config can change this logic. When the following lines are added to the config:

1. `MyGenericClass<_>` => It will be stable, regardless of the `T`
2. `MyGenericClass<*>` => It will be stable, but only if the `T` is stable.

The issue with 2 is that `T` could also be a stable class determined by the compiler (for example `@Stable` annotated class, or a predefined stable class such as `String`). This means, that gradle plugin cannot process the stability file on its own. It needs to be done by the compiler.

How do you suggest we proceed here? Do we just silently ignore `MyGenericClass<*>` syntax and treat it as `MyGenericClass<_>`? This could lead to false negatives, but I think it would be better than not supporting the configuration file at all, as this generic syntax appears to be rare (it is not supported in the IDEA plugin and nobody seems to be complaining). Alternatively, this would need to be rewritten to make these checks in the compiler.

File: stability-gradle/src/main/kotlin/com/skydoves/compose/stability/gradle/StabilityConfigParser.kt [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+  companion object {
+    private const val PATTERN_SINGLE_WILD = "\\w+"
+    private const val PATTERN_MULTI_WILD = "[\\w\\.]+"
+    private const val PATTERN_PACKAGE_SEGMENT = "\\."
+
+    private val validPatternMatcher =
+      Regex(
+        "((\\w+\\*{0,2}|\\*{1,2})\\.)*" +
+          "((\\w+(<?(?<genericmask>([*|_],)*[*|_])>)+)|(\\w+\\*{0,2}|\\*{1,2}))",
+      )
```
[the remark this answers is not in the document]
Comment (matejdro) [reply stability-gradle/src/main/kotlin/com/skydoves/compose/stability/gradle/StabilityConfigParser.kt:234]: I would not change this, it's directly from Google code, so even if it has bugs, the bugs are the same as the Google's native parser

File: stability-gradle/src/main/kotlin/com/skydoves/compose/stability/gradle/StabilityConfigParser.kt [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+    } else {
+      mask = genericMask.value
+        .split(STABILITY_GENERIC_SEPARATOR)
+        .map { if (it == STABILITY_GENERIC_INCLUDE) 1 else 0 }
+        .reduceIndexed { i, acc, flag ->
+          acc or (flag shl i)
+        }
+
+      key = keyBuilder.subSequence(0, genericMask.range.first - 1).toString()
+    }
```
[the remark this answers is not in the document]
Comment (matejdro) [reply stability-gradle/src/main/kotlin/com/skydoves/compose/stability/gradle/StabilityConfigParser.kt:191]: I would not change this, it's directly from Google code, so even if it has bugs, the bugs are the same as the Google's native parser

**Comment by matejdro** [comment]
@skydoves what do we need to do to get this merged?

**Comment by skydoves** [comment]
Hey @matejdro, sorry for the late response. I will review this as soon as possible.
